encoding/inputdecoder/decoder.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package inputdecoder
+
+import (
+	"bufio"
+	"encoding/json"
+	"io"
+
+	gyaml "github.com/ghodss/yaml"
+	"gopkg.in/yaml.v2"
+)
+
+// Decoder interface exposes method for reading multiple input requests
+type Decoder interface {
+	// NextJSONBytes returns JSON marshaled bytes of the next request body
+	// It returns io.EOF when there are no more requests
+	NextJSONBytes() ([]byte, error)
```

Comment:
Why are we using JSON here? The current serializer interface uses YAML as far as I can tell (and the specific encoding converts YAML to JSON if required, e.g., for Proto)
https://github.com/yarpc/yab/blob/dev/encoding/protobuf.go#L110

Are we planning to change the `Serializer` interface to work with JSON? If not, then we probably should use YAML here as well for consistency, and to avoid unnecessarily converting YAML to JSON if the underlying encoder (E.g., thrift) can work fine with YAML types.

YAML may allow us to express types that JSON cannot, so the conversion may not be lossless.

Reply:
`Serializer` will continue to use YAML input, I thought of Protobuf encoder while coming up with input-decoder interface and used JSON since they eventually convert YAML to JSON like you pointed out. 

I agree with the lossless conversion point, but given this decoder would be used only with protobuf encoder for streaming requests can we assume it would be safe to have this interface?

Reply:
True that it'll only be used in protobuf, but since this is defined as a more general interface, I think it's preferred to use YAML here (for consistency), and push the lossy conversion closer to where it's required. I'd like to convert YAML to protobuf directly too one day..

Reply:
Addressed this in #321